#  T-tests and P-values

In statistics, t-test is used to test if two data samples have a significant difference between their means. There are two types of t-test:

* **Student's t-test** (a.k.a. independent or uncorrelated t-test). This type of t-test is to compare the samples of **two independent populations** (e.g. test scores of students in two different classes). `scipy` provides the [`ttest_ind`](https://docs.scipy.org/doc/scipy-0.15.1/reference/generated/scipy.stats.ttest_ind.html) method to conduct student's t-test.

* **Paired t-test** (a.k.a. dependent or correlated t-test). This type of t-test is to compare the samples of **the same population** (e.g. scores of different tests of students in the same class). `scipy` provides the [`ttest_re`](https://docs.scipy.org/doc/scipy-0.15.1/reference/generated/scipy.stats.ttest_rel.html) method to conduct paired t-test.

Both types of t-tests return a number which is called the **p-value**. If p-value is below 0.05, we can confidently declare the null-hypothesis is rejected and the difference is significant. If p-value is between 0.05 and 0.1, we may also declare the null-hypothesis is rejected but we are not highly confident. If p-value is above 0.1 we do not reject the null-hypothesis.

Read more about the t-test in [this article](http://b.link/test50) and [this Quora](http://b.link/unpaired97). Make sure you understand when to use which type of t-test. 

![logo_ironhack_blue 7](https://user-images.githubusercontent.com/23629340/40541063-a07a0a8a-601a-11e8-91b5-2f13e4e6b441.png)

# Lab | Inferential statistics - T-test & P-value

### Instructions

**1. *One tailed t-test* - In a packing plant, a machine packs cartons with jars. It is supposed that a new machine will pack faster on the average than the machine currently used. To test that hypothesis, the times it takes each machine to pack ten cartons are recorded. The results, in seconds, are shown in the tables in the file `files_for_lab/machine.txt`.
   Assume that there is sufficient evidence to conduct the t test, does the data provide sufficient evidence to show if one machine is better than the other?**

In [1]:
# Import libraries
import pandas as pd
import numpy as np
import scipy.stats as st

In [2]:
##############Hypotheses############

#H0 - The new machine's packing time is the same or more, on average
#H1 - The new machine's packing time is faster, on average

In [25]:
# import data
data = pd.read_csv('machine2.txt', sep = ' ')
data

New_machine  Old_machine
0         42.1         42.7
1         41.0         43.6
2         41.3         43.8
3         41.8         43.3
4         42.4         42.5
5         42.8         43.5
6         43.2         43.1
7         42.3         41.7
8         41.8         44.0
9         42.7         44.1

In [27]:
print("new machine mean is: ",round(data['New_machine'].mean(),2))
print("old machine mean is: ",round(data['Old_machine'].mean(),2))

new machine mean is:  42.14
old machine mean is:  43.23


In [28]:
print('p value (single tailed): ',st.ttest_1samp(data['New_machine'],43.23)[1]/2)

p value (single tailed):  0.0003483188038379669


The 0 Hypothesis is a lot less than 0.05, H0 is rejected.


**2. *Matched Pairs Test* - In this challenge we will compare dependent samples of data describing our Pokemon (file `files_for_lab/pokemon.csv`). Our goal is to see whether there is a significant difference between each Pokemon's defense and attack scores. Our hypothesis is that the defense and attack scores are equal. Compare the two columns to see if there is a statistically significant difference between them and comment your result.**

In [29]:
##############Hypotheses############

#H0 - Defense and attack scores are the same
#H1 - Defense and attack scores are different

In [38]:
pokemon = pd.read_csv('pokemon.csv')
pokemon = pokemon.drop(columns='#')
pokemon

Name   Type 1  Type 2  Total  HP  Attack  Defense  \
0                Bulbasaur    Grass  Poison    318  45      49       49   
1                  Ivysaur    Grass  Poison    405  60      62       63   
2                 Venusaur    Grass  Poison    525  80      82       83   
3    VenusaurMega Venusaur    Grass  Poison    625  80     100      123   
4               Charmander     Fire     NaN    309  39      52       43   
..                     ...      ...     ...    ...  ..     ...      ...   
795                Diancie     Rock   Fairy    600  50     100      150   
796    DiancieMega Diancie     Rock   Fairy    700  50     160      110   
797    HoopaHoopa Confined  Psychic   Ghost    600  80     110       60   
798     HoopaHoopa Unbound  Psychic    Dark    680  80     160       60   
799              Volcanion     Fire   Water    600  80     110      120   

     Sp. Atk  Sp. Def  Speed  Generation  Legendary  
0         65       65     45           1      False  
1         80       80     60           1      False  
2        100      100     80           1      False  
3        122      120     80           1      False  
4         60       50     65           1      False  
..       ...      ...    ...         ...        ...  
795      100      150     50           6       True  
796      160      110    110           6       True  
797      150      130     70           6       True  
798      170      130     80           6       True  
799      130       90     70           6       True  

[800 rows x 12 columns]

In [39]:
st.ttest_rel(pokemon['Attack'], pokemon['Defense'])

Ttest_relResult(statistic=4.325566393330478, pvalue=1.7140303479358558e-05)

In [ ]:
#E-05, the H1 is alive! We reject H0 again, as the pvalue < 0.05 

In [40]:
# Import libraries
import pandas as pd
import numpy as np
import scipy.stats as st
import matplotlib.pyplot as plt
import seaborn as sns

#### Import dataset

In this challenge we will work on the Pokemon dataset you have already used. The goal is to test whether different groups of pokemon (e.g. Legendary vs Normal, Generation 1 vs 2, single-type vs dual-type) have different stats (e.g. HP, Attack, Defense, etc.). Use Ironhack's database to load the data (db: pokemon, table: pokemon_stats). 

In [51]:
pokemon = pd.read_csv('pokemon.csv')
pokemon = pokemon.drop(columns='#')
pokemon

Name   Type 1  Type 2  Total  HP  Attack  Defense  \
0                Bulbasaur    Grass  Poison    318  45      49       49   
1                  Ivysaur    Grass  Poison    405  60      62       63   
2                 Venusaur    Grass  Poison    525  80      82       83   
3    VenusaurMega Venusaur    Grass  Poison    625  80     100      123   
4               Charmander     Fire     NaN    309  39      52       43   
..                     ...      ...     ...    ...  ..     ...      ...   
795                Diancie     Rock   Fairy    600  50     100      150   
796    DiancieMega Diancie     Rock   Fairy    700  50     160      110   
797    HoopaHoopa Confined  Psychic   Ghost    600  80     110       60   
798     HoopaHoopa Unbound  Psychic    Dark    680  80     160       60   
799              Volcanion     Fire   Water    600  80     110      120   

     Sp. Atk  Sp. Def  Speed  Generation  Legendary  
0         65       65     45           1      False  
1         80       80     60           1      False  
2        100      100     80           1      False  
3        122      120     80           1      False  
4         60       50     65           1      False  
..       ...      ...    ...         ...        ...  
795      100      150     50           6       True  
796      160      110    110           6       True  
797      150      130     70           6       True  
798      170      130     80           6       True  
799      130       90     70           6       True  

[800 rows x 12 columns]

#### First we want to define a function with which we can test the means of a feature set of two samples. 

In the next cell you'll see the annotations of the Python function that explains what this function does and its arguments and returned value. This type of annotation is called **docstring** which is a convention used among Python developers. The docstring convention allows developers to write consistent tech documentations for their codes so that others can read. It also allows some websites to automatically parse the docstrings and display user-friendly documentations.

Follow the specifications of the docstring and complete the function.

In [52]:
legendary = pokemon.loc[pokemon['Legendary'] == True]
normal = pokemon.loc[pokemon['Legendary'] == False]

In [53]:
features=['HP', 'Attack', 'Defense', 'Sp. Atk', 'Sp. Def', 'Speed', 'Total']

In [54]:
s_1 = legendary[features]
s_2 = legendary[features]

In [55]:
def t_test_features(s_1, s_2, features=['HP', 'Attack', 'Defense', 'Sp. Atk', 'Sp. Def', 'Speed', 'Total']):
    """Test means of a feature set of two samples
   
    Args:
        s1 (dataframe): sample 1
        s2 (dataframe): sample 2
        features (list): an array of features to test
    Returns:
        dict: a dictionary of t-test scores for each feature where the feature name is the key and the p-value is the value
    """

    # Your code here
    results = {}
    for feature in features:
        results[feature] = st.ttest_ind(s_1[feature], s_2[feature], equal_var = False).pvalue
    return results



#### Using the `t_test_features` function, conduct t-test for Lengendary vs non-Legendary pokemons.

*Hint: your output should look like below:*

```
{'HP': 1.0026911708035284e-13,
 'Attack': 2.520372449236646e-16,
 'Defense': 4.8269984949193316e-11,
 'Sp. Atk': 1.5514614112239812e-21,
 'Sp. Def': 2.2949327864052826e-15,
 'Speed': 1.049016311882451e-18,
 'Total': 9.357954335957446e-47}
 ```

In [56]:
t_test_features(s_1, s_2)

{'HP': 1.0,
 'Attack': 1.0,
 'Defense': 1.0,
 'Sp. Atk': 1.0,
 'Sp. Def': 1.0,
 'Speed': 1.0,
 'Total': 1.0}

#### From the test results above, what conclusion can you make? Do Legendary and non-Legendary pokemons have significantly different stats on each feature?

No, all the p-values are << (much lower) than 0.05 so we can conclude the variances of the 2 samples are really small

They dont have significantly different stats between LEGENDARY and NORMAL

#### Next, conduct t-test for Generation 1 and Generation 2 pokemons.

In [57]:
# Your code here
type_1 = pokemon.loc[pokemon['Generation'] == 1]

type_2 = pokemon.loc[pokemon['Generation'] == 2]

t_test_features(type_1, type_2)

{'HP': 0.14551697834219623,
 'Attack': 0.24721958967217725,
 'Defense': 0.5677711011725426,
 'Sp. Atk': 0.12332165977104394,
 'Sp. Def': 0.18829872292645752,
 'Speed': 0.00239265937312135,
 'Total': 0.5631377907941676}

#### What conclusions can you make?

No, all the p-values are << (much lower) than 0.05 so we can conclude the variances of the 2 samples are really small

They dont have significantly different stats between GENERATION1 and GENERAIION2 

#### Compare pokemons who have single type vs those having two types.

In [58]:
# Your code here
single_type = pokemon.loc[pokemon['Type 2'].isnull()]

dual_type = pokemon.loc[pokemon['Type 2'].notnull()]

t_test_features(single_type, dual_type)


{'HP': 0.11314389855379421,
 'Attack': 0.00014932578145948305,
 'Defense': 2.7978540411514693e-08,
 'Sp. Atk': 0.00013876216585667842,
 'Sp. Def': 0.00010730610934512779,
 'Speed': 0.02421703281819094,
 'Total': 1.1157056505229961e-07}

#### What conclusions can you make?

No, all the p-values are << (much lower) than 0.05 so we can conclude the variances of the 2 samples are really small

They dont have significantly different stats between GENERATION1 and GENERATION2 

#### Now, we want to compare whether there are significant differences of `Attack` vs `Defense`  and  `Sp. Atk` vs `Sp. Def` of all pokemons. Please write your code below.

*Hint: are you comparing different populations or the same population?*

#### What conclusions can you make?

In [60]:
# H0: no relevant difference between Attack and Defense
# H1: relevant difference 
print(st.ttest_rel(pokemon.Attack, pokemon.Defense))

Ttest_relResult(statistic=4.325566393330478, pvalue=1.7140303479358558e-05)


**p-value=1.71 > 0.05 => We do not reject H0, NO RELEVANT DIFFERENCE BETWEEN Attack and Defense**

In [50]:
# H0: no relevant difference between Sp. Atk and Sp. Def
# H1: relevant difference 

print(st.ttest_rel(pokemon['Sp. Atk'], pokemon['Sp. Def']))

Ttest_relResult(statistic=0.853986188453353, pvalue=0.3933685997548122)


**p-value=0.3933 > 0.05 => We reject H0, THERE IS A RELEVANT DIFFERENCE BETWEEN Sp. Atk and Sp. Def**